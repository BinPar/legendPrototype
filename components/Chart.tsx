import React, { useState, useEffect } from 'react';
import { useRouter } from 'next/router';
import Filter from './Filter';
import useWindowSize from '../hooks/useWindowSize';
import { ChartProps, LabelInfo } from '../model/chart';
import generateRandomLegends from '../utils/generateRandomLegends';
import InterestPoint from './InterestPoint';
import {
  marcCrossingLines,
  haveCrossingLines,
} from '../utils/detectCrossingLines';
import removeCrossingLines from '../utils/removeCrossingLines';

/**
 * Generate the chart calculating the legend positions with random
 * points and legends using a deterministic Seed
 * @param cartProps Properties of the chart
 */
const Chart = ({ seed }: ChartProps): JSX.Element => {
  const size = useWindowSize();
  const router = useRouter();
  const [interestPoints, setInterestPoints] = useState<LabelInfo[]>(
    new Array<LabelInfo>(),
  );

  if (!size.width) {
    return null;
  }

  useEffect((): void => {
    let result = generateRandomLegends(seed, size);
    if (haveCrossingLines(result)) {
      result = removeCrossingLines(result);
      result = marcCrossingLines(result);
    }
    setInterestPoints(result);
  }, [seed, size]);

  /**
   * Generates a new random URL when we click anywhere in the SVG
   * using a random 7 digits number (or smaller) a the seed
   */
  const update = (): void => {
    let newSeed = seed;

    // There is almost no option, but we asure that
    // we do not repeat the last seed
    do {
      newSeed = `${Math.floor(Math.random() * 10000000)}`;
    } while (newSeed === seed);

    router.push('/sample/[seed]', `/sample/${newSeed}`, { shallow: true });
  };

  if (interestPoints.length === 0) {
    return <div className="view" />;
  }

  return (
    <div className="view">
      <svg width={size.width} height={size.height} onClick={update}>
        <defs>
          <Filter />
        </defs>
        {interestPoints.map(InterestPoint)}
      </svg>
      {interestPoints.map((label) => (
        <div
          key={label.key}
          className={`label ${label.labelPosition}`}
          style={{
            top: label.ty,
            width: label.labelWidth - 15,
            left: label.labelPosition === 'left' ? 0 : label.tx,
          }}
        >
          {label.name}
        </div>
      ))}
    </div>
  );
};

export default Chart;
