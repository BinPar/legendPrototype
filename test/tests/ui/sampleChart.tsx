import React from 'react';
import { mount } from 'enzyme';
import mockNextUseRouter from '../../mocks/mockNextUseRouter';
import SeedPage from '../../../pages/sample/[seed]';
import Chart from '../../../components/Chart';

/**
 * Checks that the sample page redirects to some other
 * sample if we clock on the SVG
 */
test('Home Page', () => {

  const seed = 480829;
  /**
   * Mocks the next router
   * with the info of the 
   * current page
   */
  mockNextUseRouter({
    route: '/sample/[seed]',
    pathname: '/sample/[seed]',
    query: {
      seed,
    },
    asPath: `/sample/${seed}`,
  });

  // Renders the page
  const seedPage = mount(<SeedPage />);

  // Get the first chart component
  const chart = seedPage.find(Chart).first();

  // Check that it have the right seed provided 
  // by the mocked router
  expect(chart.props().seed).toBe(seed.toString());
  
  // Verify that the Virtual DOM is not changed
  expect(chart).toMatchSnapshot();

  // Unmount the component
  seedPage.unmount();

});
