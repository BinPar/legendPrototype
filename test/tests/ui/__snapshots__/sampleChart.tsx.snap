// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page 1`] = `
<Chart
  seed="480829"
>
  <div
    className="view"
  >
    <svg
      height={768}
      onClick={[Function]}
      width={1024}
    >
      <defs>
        <Filter>
          <filter
            height="500%"
            id="sofGlow"
            width="500%"
            x="-200%"
            y="-200%"
          >
            <feMorphology
              in="SourceAlpha"
              operator="dilate"
              radius="1"
              result="thicken"
            />
            <feGaussianBlur
              in="thicken"
              result="blurred"
              stdDeviation="2"
            />
            <feFlood
              floodColor="rgb(255,255,255)"
              result="glowColor"
            />
            <feComposite
              in="glowColor"
              in2="blurred"
              operator="in"
              result="softGlow_colored"
            />
            <feMerge>
              <feMergeNode
                in="softGlow_colored"
              />
              <feMergeNode
                in="SourceGraphic"
              />
            </feMerge>
          </filter>
        </Filter>
      </defs>
      <g
        filter="url(#sofGlow)"
        key="^mzAmh_\`0\\""
      >
        <circle
          cx={367.8}
          cy={501.6}
          fill="#000"
          filter="url(#sofGlow)"
          key="^mzAmh_\`0\\""
          name="Commodo Lorem consequat elit"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={367.8}
          x2={512}
          y1={501.6}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key="]wwbXgrOUN"
      >
        <circle
          cx={698.8}
          cy={338.6}
          fill="#000"
          filter="url(#sofGlow)"
          key="]wwbXgrOUN"
          name="Quis consectetur nisi Lorem"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={698.8}
          x2={512}
          y1={338.6}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key="wFb@u6CoW["
      >
        <circle
          cx={676.8}
          cy={157.60000000000002}
          fill="#000"
          filter="url(#sofGlow)"
          key="wFb@u6CoW["
          name="Lorem labore magna"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={676.8}
          x2={512}
          y1={157.60000000000002}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key="\\"ruq(0#=nn"
      >
        <circle
          cx={589.8}
          cy={165.60000000000002}
          fill="#000"
          filter="url(#sofGlow)"
          key="\\"ruq(0#=nn"
          name="Velit non sunt"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={589.8}
          x2={512}
          y1={165.60000000000002}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key="D26eR]EfPd"
      >
        <circle
          cx={260.8}
          cy={548.6}
          fill="#000"
          filter="url(#sofGlow)"
          key="D26eR]EfPd"
          name="Velit reprehenderit sunt adipisicing"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={260.8}
          x2={512}
          y1={548.6}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key="AL{;[T&x4'"
      >
        <circle
          cx={625.8}
          cy={392.6}
          fill="#000"
          filter="url(#sofGlow)"
          key="AL{;[T&x4'"
          name="Amet"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={625.8}
          x2={512}
          y1={392.6}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key="\`2Rd>&?ZD%"
      >
        <circle
          cx={532.8}
          cy={255.60000000000002}
          fill="#000"
          filter="url(#sofGlow)"
          key="\`2Rd>&?ZD%"
          name="Amet pariatur est officia"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={532.8}
          x2={512}
          y1={255.60000000000002}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key="1uq/Mq7Fb*"
      >
        <circle
          cx={522.8}
          cy={366.6}
          fill="#000"
          filter="url(#sofGlow)"
          key="1uq/Mq7Fb*"
          name="Do"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={522.8}
          x2={512}
          y1={366.6}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key="q\\"x1d\\"tM/q"
      >
        <circle
          cx={211.8}
          cy={593.6}
          fill="#000"
          filter="url(#sofGlow)"
          key="q\\"x1d\\"tM/q"
          name="Fugiat laborum nostrud"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={211.8}
          x2={512}
          y1={593.6}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key="eaO;Z]lN\\"\\""
      >
        <circle
          cx={673.8}
          cy={430.6}
          fill="#000"
          filter="url(#sofGlow)"
          key="eaO;Z]lN\\"\\""
          name="Incididunt ipsum Lorem"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={673.8}
          x2={512}
          y1={430.6}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key="b(\`QS\\\\gF7j"
      >
        <circle
          cx={333.8}
          cy={391.6}
          fill="#000"
          filter="url(#sofGlow)"
          key="b(\`QS\\\\gF7j"
          name="Amet commodo"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={333.8}
          x2={512}
          y1={391.6}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key=";n$\`WKxF[6"
      >
        <circle
          cx={683.8}
          cy={179.60000000000002}
          fill="#000"
          filter="url(#sofGlow)"
          key=";n$\`WKxF[6"
          name="Deserunt deserunt consequat sunt"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={683.8}
          x2={512}
          y1={179.60000000000002}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key="y[Nef|fyiV"
      >
        <circle
          cx={463.8}
          cy={226.60000000000002}
          fill="#000"
          filter="url(#sofGlow)"
          key="y[Nef|fyiV"
          name="Velit ea"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={463.8}
          x2={512}
          y1={226.60000000000002}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key=";ctqH>%oim"
      >
        <circle
          cx={354.8}
          cy={458.6}
          fill="#000"
          filter="url(#sofGlow)"
          key=";ctqH>%oim"
          name="Excepteur ullamco minim cillum"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={354.8}
          x2={512}
          y1={458.6}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key=")'Z9kP(0*5"
      >
        <circle
          cx={509.8}
          cy={365.6}
          fill="#000"
          filter="url(#sofGlow)"
          key=")'Z9kP(0*5"
          name="Quis nostrud"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={509.8}
          x2={512}
          y1={365.6}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key="9.\\\\pwcsW'\\\\"
      >
        <circle
          cx={462.8}
          cy={251.60000000000002}
          fill="#000"
          filter="url(#sofGlow)"
          key="9.\\\\pwcsW'\\\\"
          name="Qui"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={462.8}
          x2={512}
          y1={251.60000000000002}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key="rF{[$3KSMv"
      >
        <circle
          cx={338.8}
          cy={444.6}
          fill="#000"
          filter="url(#sofGlow)"
          key="rF{[$3KSMv"
          name="Ex est commodo magna"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={338.8}
          x2={512}
          y1={444.6}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key="F4h.it<?\\\\\`"
      >
        <circle
          cx={781.8}
          cy={208.60000000000002}
          fill="#000"
          filter="url(#sofGlow)"
          key="F4h.it<?\\\\\`"
          name="Consequat qui"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={781.8}
          x2={512}
          y1={208.60000000000002}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key="\\\\[NwNgr{]w"
      >
        <circle
          cx={578.8}
          cy={335.6}
          fill="#000"
          filter="url(#sofGlow)"
          key="\\\\[NwNgr{]w"
          name="Est sint deserunt"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={578.8}
          x2={512}
          y1={335.6}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key="YbO.^=<%1!"
      >
        <circle
          cx={419.8}
          cy={300.6}
          fill="#000"
          filter="url(#sofGlow)"
          key="YbO.^=<%1!"
          name="Cupidatat dolor eu tempor"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={419.8}
          x2={512}
          y1={300.6}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key="is3r;.O67b"
      >
        <circle
          cx={758.8}
          cy={170.60000000000002}
          fill="#000"
          filter="url(#sofGlow)"
          key="is3r;.O67b"
          name="Proident"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={758.8}
          x2={512}
          y1={170.60000000000002}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key="s6<Zd\\\\\\\\,B*"
      >
        <circle
          cx={554.8}
          cy={533.6}
          fill="#000"
          filter="url(#sofGlow)"
          key="s6<Zd\\\\\\\\,B*"
          name="Magna consequat"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={554.8}
          x2={512}
          y1={533.6}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key="\`bQA:khP9S"
      >
        <circle
          cx={676.8}
          cy={218.60000000000002}
          fill="#000"
          filter="url(#sofGlow)"
          key="\`bQA:khP9S"
          name="Adipisicing commodo velit"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={676.8}
          x2={512}
          y1={218.60000000000002}
          y2={384}
        />
      </g>
      <g
        filter="url(#sofGlow)"
        key="hN7-?vbJ]6"
      >
        <circle
          cx={565.8}
          cy={298.6}
          fill="#000"
          filter="url(#sofGlow)"
          key="hN7-?vbJ]6"
          name="Deserunt sunt consequat reprehenderit"
          r={5}
          tx={512}
          ty={384}
        />
        <line
          stroke="black"
          strokeWidth="3"
          x1={565.8}
          x2={512}
          y1={298.6}
          y2={384}
        />
      </g>
    </svg>
  </div>
</Chart>
`;
